Checking server user permissions to use sudo

lunardunno opened the pull request:
Adding support for the wheel group to work with the sudo package.

Checking and redefining the locale to C if the locale differs from en_US.UTF-8|C.UTF-8|C, for correct operation of the script with the server controller.

Checking that the sudo package is pre-installed if the user is not the root.

Checking the presence of the user or the required group in the sudoers file.

Checking permission to run the package manager via sudo.

Checking for no password prompt.

Implementation of using the user's home directory name to define the CUR_USER variable when whoami is missing.

Correction of descriptions and translations of errors: ServerUserNotInSudo and ServerPacketManagerError.

Checking the availability of the server user's home directory. Lack of such access leads to the incorrect information that the package manager is busy and Error 206 (ServerPacketManagerError). Or leads to the incorrect information that the server port already allocated  (Error 201)

lunardunno committed:
Username if whoami returns an error

Сommand to use home directory name if whoami returns error or is missing for prepare_host.sh.

lunardunno committed:
Username if whoami returns an error

Сommand to use home directory name if whoami returns error or is missing for prepare_host.sh.

lunardunno committed:
Update check_user_in_sudo.sh

Сommand to use home directory name if whoami returns error or is missing for check_user_in_sudo.sh.
Checking server user permissions to use sudo using a package manager or using uname.
Сhecking and redefining the system language.
Checking requirements for sudo users or root in script.

lunardunno committed:
Update check_user_in_sudo.sh

Сommand to use home directory name if whoami returns error or is missing for check_user_in_sudo.sh.
Checking server user permissions to use sudo using a package manager or using uname.
Сhecking and redefining the system language.
Checking requirements for sudo users or root in script.

lunardunno committed:
Cases have been changed and added.

Changed description of the “Server User Not In Sudo” case.
Corrected the name and description of the "ServerPacketManagerError" case. Packet to Package.
Adding a "SudoPackageIsNotPreinstalled" case.
Adding a "ServerUserNotAllowedInSudoers" case.
Adding a "ServerUserPasswordRequired" case.

lunardunno committed:
Cases have been changed and added.

Changed description of the “Server User Not In Sudo” case.
Corrected the name and description of the "ServerPacketManagerError" case. Packet to Package.
Adding a "SudoPackageIsNotPreinstalled" case.
Adding a "ServerUserNotAllowedInSudoers" case.
Adding a "ServerUserPasswordRequired" case.

lunardunno committed:
Serves errors have been changed and added.

Corrected the name of the "ServerPacketManagerError" error to "ServerPackageManagerError".
Adding a "SudoPackageIsNotPreinstalled" error.
Adding a "ServerUserNotAllowedInSudoers" error.
Adding a "ServerUserPasswordRequired" error.

lunardunno committed:
Serves errors have been changed and added.

Corrected the name of the "ServerPacketManagerError" error to "ServerPackageManagerError".
Adding a "SudoPackageIsNotPreinstalled" error.
Adding a "ServerUserNotAllowedInSudoers" error.
Adding a "ServerUserPasswordRequired" error.

lunardunno committed:
Return ServerPacketManagerError

Return to the name "ServerPacketManagerError".

lunardunno committed:
Return ServerPacketManagerError

Return to the name "ServerPacketManagerError".

lunardunno committed:
Added errors handling 

Added new errors' handling to serverController.cpp.
Permission checks are also performed for the root user.

lunardunno committed:
Added errors handling 

Added new errors' handling to serverController.cpp.
Permission checks are also performed for the root user.

lunardunno committed:
Update translations

Updating translations for two existing server errors.

lunardunno committed:
Update translations

Updating translations for two existing server errors.

lunardunno committed:
Merge branch 'sudo_permissions' of https://github.com/amnezia-vpn/amnezia-client into sudo_permissions

lunardunno committed:
Myanmar translation update

lunardunno committed:
Myanmar translation update

lunardunno committed:
Update for my_MM.ts

lunardunno committed:
Update for my_MM.ts

lunardunno committed:
checking for not allowed

Checking for "not allowed" in stdOut

lunardunno committed:
checking for not allowed

Checking for "not allowed" in stdOut

lunardunno committed:
Removed "not allowed"

Removed check for "not allowed" in stdOut

lunardunno committed:
Removed "not allowed"

Removed check for "not allowed" in stdOut

lunardunno committed:
Removed nested launch

Removed nested launch via sudo

lunardunno committed:
Removed nested launch

Removed nested launch via sudo

lunardunno committed:
Returned nested launch

Returned nested launch via sudo

lunardunno committed:
Returned nested launch

Returned nested launch via sudo

lunardunno committed:
All checks with sudo

Both checks with sudo always run.

lunardunno committed:
All checks with sudo

Both checks with sudo always run.

lunardunno committed:
Moved removing timestamp sudo

Removing the sudo timestamp is done every time.

lunardunno committed:
Moved removing timestamp sudo

Removing the sudo timestamp is done every time.

lunardunno committed:
Checking the user directory

Checking the accessibility of the user's home directory

lunardunno committed:
Checking the user directory

Checking the accessibility of the user's home directory

lunardunno committed:
Polishing

Изменение порядка обработки ошибок.

lunardunno committed:
Polishing

Изменение порядка обработки ошибок.

vanyasem commented on client/server_scripts/check_user_in_sudo.sh:
Shouldn't this script have a [shebang](https://en.wikipedia.org/wiki/Shebang_%28Unix%29) `#!/usr/bin/env bash` at the top?

lunardunno replied:
For what purposes?
This script is executed in sh.
Have you tried running it in sh?

vanyasem replied:
You're right, a shebang won't help, as the script is executed line-by-line. My mistake

However, it uses the default user's shell instead of directly calling `sh`:

1) The script is ran in `ServerController::isUserInSudo`:
https://github.com/amnezia-vpn/amnezia-client/blob/157064763e0897e60fd9461bf0635ced38315a6d/client/core/controllers/serverController.cpp#L771

2) `ServerController::runScript` sends the command to an ssh client:
https://github.com/amnezia-vpn/amnezia-client/blob/157064763e0897e60fd9461bf0635ced38315a6d/client/core/controllers/serverController.cpp#L87

3) Finally, `Client::executeCommand` calls `libssh`:
https://github.com/amnezia-vpn/amnezia-client/blob/678bfffe492ed89911a9dcacfb125e82aa1ce1be/client/core/sshclient.cpp#L148

4) `ssh_channel_request_exec` _Run a shell command without an interactive shell._: https://api.libssh.org/master/group__libssh__channel.html#ga567d509183ade0a77190f390e2b5747d

You might get confused by the phrase `This is similar to 'sh -c command'.`

This implies the fact that a non-interactive shell is spawned, similar to how `sh -c` executes commands in a non-interactive shell. It does not mean that the commands are in fact executed in `sh`. Any shell has a non-interactive mode, and the default shell for the particular user is used to run the command.

There is no guarantee that the user's shell will be POSIX-compliant. For example, having `fish` or `pwsh` set as a login shell breaks the execution of the script. See https://github.com/amnezia-vpn/amnezia-client/issues/1201

This could be fixed by prefixing every line with `/bin/sh -c` or something similar.

At the very least, there could be a check that verifies that a supported (POSIX-compliant) shell is being used by the specified ssh user.

vanyasem replied:
I actually made a small example to prove that your assumption that the code is ran in `/bin/sh` is wrong: https://gist.github.com/vanyasem/4c931f186c48da7b98b0f5518f1c6c64

`ssh_example` establishes an SSH connection via a private key, and then executes `ssh_channel_request_exec(channel, "echo $0")`. The output of the command is printed to console. `echo $0` returns the name of the currently running program, which in this case would be the name of the shell.

As you can see, changing the user's shell by calling `chsh` results in `$SHELL` being changed. If the commands were executed in `sh`, the value would always be `sh`

<img width="769" alt="Screenshot 2025-03-04 at 12 04 52" src="https://github.com/user-attachments/assets/8b8cb8f6-c6f3-4a19-8e15-4203e0acb0a8" />

vanyasem commented on client/server_scripts/prepare_host.sh:
Shouldn't this script have a [shebang](https://en.wikipedia.org/wiki/Shebang_%28Unix%29) `#!/usr/bin/env bash` at the top?

lunardunno committed:
changing detection order 

change the order of detection of inconsistencies:
1. sudo not preinstalled. (if user != root)
2. user not in sudo or wheel group. (if user != root)
3. user's directory is not accessible. (for all)
4. user not allowed in sudoers. (for all)
5. user password required. (for all)

lunardunno committed:
changing detection order 

change the order of detection of inconsistencies:
1. sudo not preinstalled. (if user != root)
2. user not in sudo or wheel group. (if user != root)
3. user's directory is not accessible. (for all)
4. user not allowed in sudoers. (for all)
5. user password required. (for all)

lunardunno committed:
Merge branch 'dev' into sudo_permissions

lunardunno committed:
Packet to Package

albexk committed:
chore: bump version (#1463)

lunardunno committed:
fix for sh (#1462)

Fix for servers where sh is used as default shell.

lunardunno committed:
Merge branch 'sudo_permissions' of https://github.com/amnezia-vpn/amnezia-client into sudo_permissions

lunardunno committed:
Undoing unintended changes

Undoing unintended changes.

lunardunno committed:
Undoing unintended change

Undoing unintended change.

lunardunno committed:
not allowed to use sudo

The user is not allowed to use sudo on this server.

lunardunno committed:
Capital letters in the error

Capital letters in the error description.
